src/helpers/formulaHelper.js

```diff
@@ -0,0 +1,187 @@
+                pushOperator(resultStack, char);
+            }
+            if (FUNCTION_TOKENS.includes(parseStack[parseStack.length - 1])) {
+                pushOperator(resultStack, parseStack.pop());
+            }
+        }
+
+        else {
+            while (parseStack.length > 0 &&
+                precedence(tokens[i]) <= precedence(parseStack[parseStack.length - 1])
+            ) {
+                pushOperator(resultStack, parseStack.pop());
+            }
+            parseStack.push(tokens[i]);
+        }
+    }
+
+    while (parseStack.length > 0) {
+        let operator = parseStack.pop();
+        pushOperator(resultStack, operator);
+    }
+
+    return resultStack;
+}
+
+function isValidNonNumberToken(char) {
+    return (
+        (char >= 'a' && char <= 'z') || 
+        (char >= 'A' && char <= 'Z') ||
+        (char == '_')
+    );
+}
+
+function precedence(operator) {
+    if (operator == '^') {
+        return 3;
+    } else if (operator == '*' || operator == '/' || operator == '÷') {
+        return 2;
+    } else if (operator == '+' || operator == '-') {
+        return 1;
+    }
+    return 0;
+}
+
+function pushOperator(resultStack, operator) {
+    // No need to return the array because array is passed by reference
+    if (FUNCTION_TOKENS.includes(operator)) {
+        resultStack.push({ token: operator, type: TOKEN_TYPE.Function });
+    } else {
+        resultStack.push({ token: operator, type: TOKEN_TYPE.Operator });
+    }
+}
+
+function evaluate(evalStack, vars) {
+    if (vars == null) {
+        throw `Missing variables`;
+    }
+
+    let resultStack = [];
+    for (let i = 0; i < evalStack.length; i++) {
+        console.log(resultStack);
+        const tokenObj = evalStack[i];
+
+        if (tokenObj.type == TOKEN_TYPE.Variable) {
+            if (!vars.hasOwnProperty(tokenObj.token)) {
+                throw `Missing variable ${tokenObj.token}`;
+            }
+            let varValue = vars[tokenObj.token];
+            resultStack.push(varValue);
+        } else if (tokenObj.type == TOKEN_TYPE.Operator) {
+            let num1 = resultStack.pop();
+            let num2 = resultStack.pop();
+            resultStack.push(applyOperator(tokenObj.token, num1, num2));
+        } else if (tokenObj.type == TOKEN_TYPE.Function) {
+            let tempArgs = [];
+            while (resultStack.length > 0) {
+                tempArgs.push(resultStack.pop());
+            }
+            resultStack.push(applyFunction(tokenObj.token, tempArgs));
+        } else {
+            resultStack.push(tokenObj.token);
+        }
+    }
+    return resultStack.pop();
+}
+
+function applyOperator(operator, num1, num2) {
+    switch (operator) {
+        case '^':
+            return Math.pow(num2, num1);
+        case '*':
+            return num2 * num1;
+        case '/':
+            return num2 / num1;
+        case '+':
+            return num2 + num1;
+        case '-':
+            return num2 - num1;
+    }
```

Comment:
Constants galore~

Reply:
Do you really need to extract literal mathematical symbols into constants? v:

Reply:
I think so

Reply:
I think it just makes the code too verbose since most people will recognize mathematical operators and what operations they will do. The instances where these mathematical operators used are also limited within `applyOperator` and `precedence` functions therefore making these into constants will have very little reusability. Are there other reasons that we should extract these into constants?

Reply:
Hm, you have a point there regarding `most people will recognize mathematical operators`

But the reason for this is not reusability actually, Clean Code practices emphasise meaningful/semantic coding. So it's not reusability but rather readability

Then I leave the decision to you :v Resolve this if you think it's better to leave it this way

Reply:
I think I will resolve this one and the one in `precedence` function. My reasoning is that converting mathematical operators into constants will only slightly improve the readability because the code already explains itself, e.g. if the input operator is `-` then apply operation `num2` minus `num1`.

For other cases like parenthesis, comma, and function names, I think converting those into functions or constants are more appropriate since some of them doesn't really explain what they do.